lib/ocpp/v201/profile.cpp

```diff
@@ -0,0 +1,684 @@
+
+            // start time(s) of the schedule
+            // the start time of this period is calculated in period_entry_t::init()
+            const auto schedule_start = calculate_start(in_now, in_end, in_session_start, in_profile);
+
+            for (std::uint8_t i = 0; i < schedule_start.size(); i++) {
+                const bool has_next_occurrance = (i + 1) < schedule_start.size();
+                const auto& entry_start = schedule_start[i];
+
+                /*
+                 * The duration of this period (from the start of the schedule) is the sooner of
+                 * - forever
+                 * - next period start time
+                 * - optional duration
+                 * - the start of the next recurrence
+                 * - optional validTo
+                 */
+
+                int duration = std::numeric_limits<int>::max(); // forever
+
+                if (has_next_period) {
+                    duration =
+                        in_profile.chargingSchedule.front().chargingSchedulePeriod[in_period_index + 1].startPeriod;
+                }
+
+                // check optional chargingSchedule duration field
+                if (in_profile.chargingSchedule.front().duration &&
+                    (in_profile.chargingSchedule.front().duration.value() < duration)) {
+                    duration = in_profile.chargingSchedule.front().duration.value();
+                }
+
+                // check duration doesn't extend into the next recurrence
+                if (has_next_occurrance) {
+                    const auto next_start =
+                        duration_cast<seconds>(schedule_start[i + 1].to_time_point() - entry_start.to_time_point())
+                            .count();
+                    if (next_start < duration) {
+                        duration = next_start;
+                    }
+                }
+
+                // check duration doesn't extend beyond profile validity
+                if (in_profile.validTo) {
+                    // note can be negative
+                    const auto valid_to = floor_seconds(in_profile.validTo.value());
+                    const auto valid_to_seconds =
+                        duration_cast<seconds>(valid_to.to_time_point() - entry_start.to_time_point()).count();
+                    if (valid_to_seconds < duration) {
+                        duration = valid_to_seconds;
+                    }
+                }
+
+                period_entry_t entry;
+                entry.init(entry_start, duration, this_period, in_profile);
+                const auto now = floor_seconds(in_now);
+
+                if (entry.validate(in_profile, now)) {
+                    entries.push_back(std::move(entry));
+                }
+            }
+        }
+    }
+
+    return entries;
+}
+
+std::vector<period_entry_t> calculate_profile(const DateTime& now, const DateTime& end,
```

Comment:
A name like `calculate_period_entries` might be clearer here.

Reply:
Agreed. Right now the idea is to hold off on refactoring until later cards, and just match the names and patterns of v1.6.